File: C++/3.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAXN = 100005;
vector<int> adj[MAXN];
int disc[MAXN], low[MAXN], timer;
bool visited[MAXN];
vector<pair<int, int>> bridges;

void dfs(int u, int parent) {
    visited[u] = true;
    disc[u] = low[u] = ++timer;

    for (int v : adj[u]) {
        if (v == parent) continue;

        if (!visited[v]) {
            dfs(v, u);
            low[u] = min(low[u], low[v]);

            if (low[v] > disc[u]) {
                bridges.push_back({min(u, v), max(u, v)});
            }
        } else {
            low[u] = min(low[u], disc[v]);
        }
    }
}

int main() {
    int V, E;
    cout << "Enter number of vertices and edges: ";
    cin >> V >> E;

    for (int i = 0; i < E; ++i) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    memset(visited, 0, sizeof(visited));
    timer = 0;

    for (int i = 0; i < V; ++i) {
        if (!visited[i]) {
            dfs(i, -1);
        }
    }

    cout << "Critical Links (Bridges):" << endl;
    sort(bridges.begin(), bridges.end());
    for (auto& b : bridges) {
        cout << b.first << " " << b.second << endl;
    }

    return 0;
}

File: Python/3.py
"""Find Bridges (Critical Links) in an Undirected Graph — Tarjan's Algorithm"""

import sys
sys.setrecursionlimit(200000)

def find_bridges(V, adj):
    disc = [0] * V
    low = [0] * V
    visited = [False] * V
    bridges = []
    timer = [0]

    def dfs(u, parent):
        visited[u] = True
        timer[0] += 1
        disc[u] = low[u] = timer[0]

        for v in adj[u]:
            if v == parent:
                continue
            if not visited[v]:
                dfs(v, u)
                low[u] = min(low[u], low[v])
                if low[v] > disc[u]:
                    bridges.append((min(u, v), max(u, v)))
            else:
                low[u] = min(low[u], disc[v])

    for i in range(V):
        if not visited[i]:
            dfs(i, -1)

    bridges.sort()
    return bridges

if __name__ == "__main__":
    V, E = map(int, input("Enter number of vertices and edges: ").split())
    adj = [[] for _ in range(V)]

    for _ in range(E):
        u, v = map(int, input().split())
        adj[u].append(v)
        adj[v].append(u)

    bridges = find_bridges(V, adj)
    print("Critical Links (Bridges):")
    for u, v in bridges:
        print(u, v)
